realm/realm-library/src/androidTest/java/io/realm/BulkInsertTests.java

```diff
@@ -0,0 +1,584 @@
+        objA_no_pk.setColumnString("A");
+        NoPrimaryKeyWithPrimaryKeyObjectRelation objB_no_pk = new NoPrimaryKeyWithPrimaryKeyObjectRelation();
+        objB_no_pk.setColumnString("B");
+        AllTypesPrimaryKey objC_pk = new AllTypesPrimaryKey();
+        objC_pk.setColumnLong(7);
+        objC_pk.setColumnString("C");
+        AllTypesPrimaryKey objD_pk = new AllTypesPrimaryKey();
+        objD_pk.setColumnLong(7);
+        objD_pk.setColumnString("D");
+
+        objA_no_pk.setColumnRealmObjectPK(objC_pk);
+        objB_no_pk.setColumnRealmObjectPK(objD_pk);
+
+        ArrayList<NoPrimaryKeyWithPrimaryKeyObjectRelation> objects = new ArrayList<NoPrimaryKeyWithPrimaryKeyObjectRelation>(2);
+        objects.add(objA_no_pk);
+        objects.add(objB_no_pk);
+
+        realm.beginTransaction();
+        realm.insertOrUpdate(objects);
+        realm.commitTransaction();
+
+        RealmResults<NoPrimaryKeyWithPrimaryKeyObjectRelation> all = realm.where(NoPrimaryKeyWithPrimaryKeyObjectRelation.class).findAllSorted("columnString", Sort.DESCENDING);
+        assertEquals(2, all.size());
+        assertEquals("B", all.get(0).getColumnString());
+        assertEquals("A", all.get(1).getColumnString());
+
+        assertNotNull(all.get(0).getColumnRealmObjectPK());
+        assertNotNull(all.get(1).getColumnRealmObjectPK());
+
+        assertEquals("D", all.get(0).getColumnRealmObjectPK().getColumnString());
+        assertEquals("D", all.get(1).getColumnRealmObjectPK().getColumnString());
+
+        assertEquals(1, realm.where(AllTypesPrimaryKey.class).findAll().size());
+    }
+
+    //any omitted argument should not end in a SIGSEGV but an exception
+
+    @Test
+    public void testObjectNull() {
+        AllTypes nullObject = null;
+
+        realm.beginTransaction();
+        try {
+            realm.insert(nullObject);
+            fail("Should trigger NullPointerException");
+        } catch (NullPointerException ignore) {
+
+        } finally {
+            realm.cancelTransaction();
+        }
+    }
+
+    @Test
+    public void testListNull() {
+        List<AllTypes> nullObjects = null;
+
+        realm.beginTransaction();
+        try {
+            realm.insert(nullObjects);
+            fail("Should trigger NullPointerException");
+        } catch (NullPointerException ignore) {
+
+        } finally {
+            realm.cancelTransaction();
+        }
+    }
+
+    @Test
+    public void testListNullElement() {
+        Dog dog1 = new Dog();
+        dog1.setName("Dog 1");
+        Dog dog2 = new Dog();
+        dog2.setName("Dog 2");
+        ArrayList<Dog> list = new ArrayList<Dog>();
+        list.addAll(Arrays.asList(dog1, null, dog2));
+
+        realm.beginTransaction();
+        try {
+            realm.insert(list);
+        } catch (NullPointerException ignore) {
+
+        } finally {
+            realm.cancelTransaction();
+        }
+    }
+
+    @Test
+    public void testManagedObject() {
```

Comment:
Add a comment about expected outcome. e.g ``//Inserting a managed object will result in it being copied.